### PR TITLE
Use Task Output as Extra for Dataset Trigger and DAG Run Conf

### LINKED ISSUE #37810: Annotate a Dataset Event in the Source Task
### Description

To eventually support the construct and UI we’re aiming for in assets, we need to attach metadata to the actual data, not the task that produces it, nor the location it is written to.

In the task-based web UI, we can show those attached metadata in the task that emits the dataset event, to give an impression the metadata is directly associated to the task. In the implementation, however, the metadata would only be associated to the dataset, and only indirectly related to the task by the fact that the task emits the event.

### Use case/motivation

An Airflow task generating data may want to attach information about it. Airflow does not currently provide a good interface for this. The only thing that resembles such feature is to attach and extra dict on `Dataset` like this:

```python
@task(outlets=[Dataset("s3://bucket/key", extra={"xx": "yy"})])
def producer():
    # Write to the file on S3...
```

This is however quite limiting. It may be good enough for static information such as *who owns this data*, but not for information that is only known at runtime, to provide additional context to the generated data.

### Store runtime-populated extras on `DatasetEvent`

When a Dataset event is emitted, the corresponding `DatasetEvent` model in the database already has a field call `extra`. This is however currently not populated when the event is generated from a task outlet (only when it’s created via the REST API).

[A previous design discussion contains the following comment](https://github.com/apache/airflow/issues/35297#issuecomment-1843720880) from @blag:

> The only intent that I ever had (and note: I was not the author of the dataset AIP) regarding the `extra` fields for datasets and dataset events, was to allow third party integrations to easily store information from external systems that wasn't captured in Airflow's database schema, eg: to do so without forking the schema migrations.

and

> But if all you are looking to do is pass information between task instances in the same DAG run or between task instances in different DAG runs, I believe the Airflow mechanism to do this is XComs, even with data-aware Airflow.

However, I would argue that user code in an Airflow DAG should also have the ability to store custom information. While the information is readable in downstream tasks—thus technically is a mechanism to pass data between tasks—the main intention behind the design is instead to annotate the generated data, and does not go against the original design.

### Provide extras at runtime

The task function (either `@task`-decorated, or a classic operator’s `execute`) will be able to attach values to a Dataset URI in the function. This is done by an accessor proxy under the key `dataset_events`, so in the task function you can:

```python
@task(outlets=[Dataset("s3://bucket/key")])
def producer(*, dataset_events: dict[str, DatasetEventProxy]):
    dataset_events["s3://bucket/key"].extra["desc"] = "foo bar"
```

After the task function’s execution, the extras provided dynamically are written to the `DatasetEvent` entry generated for the Dataset. Do note specially this is entirely distinct from `extra` on `Dataset`.

Instead of using URI, you can use the dataset object directly to access the proxy:

```python
target = Dataset("s3://bucket/key")

@task(outlets=[target])
def producer(*, dataset_events: dict[str, DatasetEventProxy]):
    dataset_events[target].extra["desc"] = "foo bar"
```

Example using the context dict instead:

```python
@task(outlets=[target])
def producer():
    context = get_current_context()
    context["dataset_events"][target].extra["desc"] = "foo bar"
```

With a classic operator:

```python
class MyOperator(BaseOperator):
    def execute(self, context: Context) -> None:
        super().execute(context)
        context["dataset_events"][target].extra["desc"] = "foo bar"
```

### Show dataset event extras in web UI

Both dataset and dataset event extras currently have zero visibility to users in the web UI. This is somewhat acceptable for datasets, where the extra dict is static, but is a problem for dynamically generated values. Additional components should be added to the web UI to display extras emitted by a dataset event.

An obvious first addition would be to add a tables in the task instance panel in the Grid view when the task instance emits dataset events with extras. Quick UI mock:

dataset-extra-view-mock

Each key and value will simply be stringified to be displayed in the table. This should be enough for simple data since the extra dict currently need to be JSON-compatible. We can discuss richer data (similar to how Jupyter displays a DataFrame), and putting this information in other places (e.g. in the Dataset view) in the future.

### Related issues

https://github.com/apache/airflow/issues/35297
https://github.com/apache/airflow/pull/36075

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by jscheffl**
This PR is a proposal to fix/resolve the request for feature as described in #37810 as a thin approach.

Idea for this solution:

- Allow using task result as dynamic `extra`
  - If dataset defined `extra_from_return=True` use task result as extra
- Pass the dict of `extra` to the data triggered DAG as configuration/parameter
  - If multiple events trigger a DAG, `extra` dictionaries are merged

This PR super-seeds PR #37888 

closes: #37810

### CODE REVIEW

File: airflow/models/taskinstance.py [hunk 1]
```diff
@@ -2553,21 +2554,30 @@ def _run_raw_task(
...
-    def _register_dataset_changes(self, *, session: Session) -> None:
+    def _register_dataset_changes(self, result: Any, *, session: Session) -> None:
         if TYPE_CHECKING:
             assert self.task
 
         for obj in self.task.outlets or []:
             self.log.debug("outlet obj %s", obj)
             # Lineage can have other types of objects besides datasets
             if isinstance(obj, Dataset):
+                if obj.extra:
```
Comment (dirrao) [inline airflow/models/taskinstance.py:2565]: Can we merge the event static information with dynamic information?

File: airflow/models/taskinstance.py [hunk 2]
```diff
@@ -2553,21 +2554,30 @@ def _run_raw_task(
...
+                    extra = obj.extra
```
Comment (uranusjr) [inline airflow/models/taskinstance.py:2566]: I do not think this is right. It’s made quite clear in previous issues that `Dataset.extras` and `DatasetEvent.extras` are different things and should be kept separate.

File: airflow/models/taskinstance.py [hunk 3]
```diff
@@ -2553,21 +2554,30 @@ def _run_raw_task(
...
+                elif obj.extra_from_return:
+                    extra = result if isinstance(result, dict) else {str(self.task_id): result}
```
Comment (uranusjr) [inline airflow/models/taskinstance.py:2568]: Automatically putting the value under a key is too magical to me. I would prefer this either just forward the value (the extra field is capable of storing any JSON-able values, after all), or skip the value entirely if it has an unexpected type. The task ID is also not a particularly obvious key either.

File: airflow/jobs/scheduler_job_runner.py [hunk 4]
```diff
@@ -1282,9 +1282,15 @@ def _create_dag_runs_dataset_triggered(
                     events=dataset_events,
                 )
 
+                run_conf = {}
+                for event in dataset_events:
+                    if event.extra:
+                        run_conf.update(event.extra)
```
Comment (uranusjr) [inline airflow/jobs/scheduler_job_runner.py:1288]: I would prefer we put this in a separate PR to discuss. It’s not entirely obvious how extras should be merged from different events that trigger the run.

File: airflow/models/taskinstance.py [hunk 5]
```diff
@@ -2455,7 +2456,7 @@ def _run_raw_task(
 
             try:
                 if not mark_success:
-                    self._execute_task_with_callbacks(context, test_mode, session=session)
+                    result = self._execute_task_with_callbacks(context, test_mode, session=session)
```
Comment (shrukul) [inline airflow/models/taskinstance.py:2459]: 1. It seems that not all operators return a `result` on execution (example: [DatabricksSubmitRunOperator](https://github.com/apache/airflow/blob/main/airflow/providers/databricks/operators/databricks.py#L563-L564))
2. Some operators do return a `result` - like ([GlueOperator](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/operators/glue.py#L217), [EmrServerlessStartJobOperator](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/operators/emr.py#L1472)), but the return value is `job_id` in these scenarios.

Questions -
1. If any of the above operators publish an Airflow dataset, how to specify `extra` dictionary in the corresponding dataset?

**Comment by uranusjr** [comment]
I gave this a longer thought and I think I’m not particularly fond of the approach mainly because it mixes `Dataset.extra` and `DatasetEvent.extra`. As described in #37810, the current design is very explicit on the two being separate—`Dataset.extra` describes the thing that’s pointed to by the URI, while `DatasetEvent.extra` describes the _data_ written into the thing represented by the URI. I quoted a few paragraphs from @blag in the issue. The design here directly passes through `Dataset.extra` to `DatasetEvent.extra`, which I believe would not make sense to most people. (For the record, I actually proposed copying `Dataset.extra` to `DatasetEvent.extra` initially, but @jedcunningham convinced me to not do it before I published #37810.)

The `extra_from_return` argument also has the same problem fundamentally—the flag controlling `DatasetEvent` behaviour should not be set on `Dataset`. It should be on the task instead (or on DatasetEvent itself, but that’s awkward because the event does not exist conceptually when we want to emit the extra).

File: airflow/models/taskinstance.py (hunk 2 above)
Comment (jscheffl) [reply airflow/models/taskinstance.py:2566]: Sorry, then it seems I mis-understood your comments in the previous PR7discussions about this. Thanks for clearly documenting the differences between `Dataset` and `DatasetEvent`(both) `extra` details in #38481. This opened my eyes and now I understand your push-back.
Now it is "clear" to me. Intent was _not_ to overwrite or mangle "static" information for events.

File: airflow/models/taskinstance.py (hunk 1 above)
Comment (jscheffl) [reply airflow/models/taskinstance.py:2565]: No we should not :-(
I (previously) interpreted the `extra`field as being dynamic data. But as it is (actually) intended to be meta data for the dataset itself I mis-interpreted this as an option to pass extra data for this "reference" to a Dataset (interpreted by the URI).

File: airflow/models/taskinstance.py (hunk 5 above)
Comment (jscheffl) [reply airflow/models/taskinstance.py:2459]: Regarding 1+2: Yes, if no result is generated, the content will be just `None`. If it is a scalar like a job_id it is rather a string. So in such cases the output is not usable for passing along to the dataset event.

Regarding your question 1: The publish of the event actually is happening _after_ this line of code. The change was attemping to capture the result to put is as extra a few lines of code later from here.

**Comment by jscheffl** [comment]
> I gave this a longer thought and I think I’m not particularly fond of the approach mainly because it mixes `Dataset.extra` and `DatasetEvent.extra`. As described in #37810, the current design is very explicit on the two being separate—`Dataset.extra` describes the thing that’s pointed to by the URI, while `DatasetEvent.extra` describes the _data_ written into the thing represented by the URI. I quoted a few paragraphs from @blag in the issue. The design here directly passes through `Dataset.extra` to `DatasetEvent.extra`, which I believe would not make sense to most people. (For the record, I actually proposed copying `Dataset.extra` to `DatasetEvent.extra` initially, but @jedcunningham convinced me to not do it before I published #37810.)
> 
> The `extra_from_return` argument also has the same problem fundamentally—the flag controlling `DatasetEvent` behaviour should not be set on `Dataset`. It should be on the task instead (or on DatasetEvent itself, but that’s awkward because the event does not exist conceptually when we want to emit the extra).

Thanks for sharing your concerns. After reading #38481 I better understand what you were meaning. I did not understand the difference between `Dataset.extra` (Which I always felt like "Dataset" being the holder for the URI as pointer towards a dataset) and the `DatasetEvent.extra` (which I did not notice before, expected this is the result of the outlet from execution not a definition made by the user). You ehancement in the docs in PR #38481 makes it now clear.

Therefore I'm now in favor of PR #38481 where you propose a simple but explicit possibility to attach extra per outlet w/o mixing with return or XCom.

Yes, and in-deed, the merging of multiple extras as DAG run conf is something separate - thought it is a 2-for-1 in this PR to publish the `extra` dynamically here as well as show how to easily consume it.

So I will close this "test baloon" PR here and "maybe" will open a separate PR for the topic of using extra as DAG run conf... if time.
